### PR TITLE
Add an example and README to the new Grafana lib

### PR DESCRIPTION
**Pull request by malcolmholmes**
and fix a few issues in the api while we're at it.

### CODE REVIEW

File: grafana/configmaps.libsonnet [hunk 1]
```diff
@@ -61,7 +61,7 @@ local configMap = k.core.v1.configMap;
     ['%s-%d' % [prefix(folder.id), shard]]+:
       configMap.new('%s-%d' % [prefix(folder.id), shard]) +
       configMap.withDataMixin({
-        [name]: std.toString(folder.dashboards[name])
+        [if std.endsWith(name, '.json') then name else '%s.json' % name]: std.toString(folder.dashboards[name])
```
Comment (Duologic) [inline grafana/configmaps.libsonnet:64]: Can you break this conditional in 3 lines? 
```
if ...
then ...
else ...
```

**Review by Duologic — APPROVED** [review]

File: grafana/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Jsonnet Library for Deploying Grafana
```
Comment (jdbaldry) [inline grafana/README.md:1]: nit: Our documentation style guide recommends sentence case

https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#headings

File: grafana/README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+## Library Notes
+### Statefulness
+Currently, this library creates a stateless Grafana. What this actually
+means is that, on startup, Grafana imports dashboard files into its
+database. However, given Grafana is installed into a deployment without
+an external volume, the database is wiped on restart. This creates a
+stateless Grafana installation.
+
+Statelessness is a **good thing** when deploying all your dashboards, etc
```
Comment (jdbaldry) [inline grafana/README.md:18]: nit: 
```suggestion
Statelessness is a **good thing** when deploying all your dashboards, etc.
```
or maybe expand the et cetera
```suggestion
Statelessness is a **good thing** when deploying all your dashboards, data sources, and plugins
```

File: grafana/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from Jsonnet code. Through it you avoid your dashboards drifting from the
+Jsonnet/version control managed known good state.
+
+A possible future extension of this library could convert the Grafana
+instance into a stateful one, mounting the database into a PVC.
+
+### Mixins
+This library does not (yet) support Monitoring Mixins directly, although
```
Comment (jdbaldry) [inline grafana/README.md:26]: Do we have a link to somewhere where we document what a "Monitoring Mixin" is?

File: grafana/dashboards.libsonnet [hunk 5]
```diff
@@ -25,11 +25,11 @@
 
   // add a new dashboard, creating the folder if necessary
   addDashboard(name, dashboard, folder=''):: {
-    local hasShards = std.objectHas(super.folders[name], 'shards'),
+    local hasShards = std.objectHas(super.grafanaDashboardFolders, folder) && std.objectHas(super.grafanaDashboardFolders[folder], 'shards'),
```
Comment (jdbaldry) [inline grafana/dashboards.libsonnet:28]: I'm not sure if this is a concern but an empty string is a valid key:
```
$ cat main.jsonnet 
local obj = {
  '': 'test',
};
$ jsonnet main.jsonnet 
{
  has: std.objectHas(obj, ''),
  value: obj[''],
  obj: obj,
}
```

Comment (malcolmholmes) [reply grafana/dashboards.libsonnet:28]: Yep, thats fine  this is just preventing breakage.

File: grafana/README.md (hunk 4 above)
Comment (malcolmholmes) [reply grafana/README.md:26]: Yep, I was lazy. Bah :-)

File: grafana/README.md (hunk 3 above)
Comment (malcolmholmes) [reply grafana/README.md:18]: And notification channels. Will add.

**Pull request merged**